File: cpp/libjoynrclustercontroller/messaging/MessagingPropertiesPersistence.cpp
/*
 * #%L
 * %%
 * Copyright (C) 2011 - 2017 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
#include "joynr/Settings.h"
#include "joynr/Util.h"

#include "MessagingPropertiesPersistence.h"

namespace joynr
{

MessagingPropertiesPersistence::MessagingPropertiesPersistence(const std::string& filename)
        : _filename(filename)
{
}

std::string MessagingPropertiesPersistence::getIdFromPersistence(const std::string& key)
{
    // Read and write to the persistence file using a Settings object
    Settings settings(_filename);

    std::string id;

    // Get the persisted specified id if one exists
    if (settings.contains(key)) {
        id = settings.get<std::string>(key);
    } else {
        // Create a new persisted id for key
        id = util::createUuid();
        settings.set(key, id);
        settings.sync();
    }

    return id;
}

std::string MessagingPropertiesPersistence::getChannelId()
{
    return getIdFromPersistence(CHANNEL_ID_KEY());
}

std::string MessagingPropertiesPersistence::getReceiverId()
{
    return getIdFromPersistence(RECEIVER_ID_KEY());
}

const std::string& MessagingPropertiesPersistence::CHANNEL_ID_KEY()
{
    static const std::string value("messaging/channelId");
    return value;
}

const std::string& MessagingPropertiesPersistence::RECEIVER_ID_KEY()
{
    static const std::string value("messaging/receiverId");
    return value;
}

} // namespace joynr

File: java/messaging/messaging-common/src/main/java/io/joynr/messaging/MessagingPropertiesPersistence.java
/*
 * #%L
 * %%
 * Copyright (C) 2011 - 2017 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
package io.joynr.messaging;

import static io.joynr.util.JoynrUtil.createUuidString;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.Properties;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import io.joynr.runtime.PropertyLoader;

public class MessagingPropertiesPersistence {
    private static final Logger logger = LoggerFactory.getLogger(MessagingPropertiesPersistence.class);
    private File persistenceFile;
    private Properties storage;

    public MessagingPropertiesPersistence(File persistenceFile) {
        this.persistenceFile = persistenceFile;
        makeDirectories(persistenceFile);
        storage = PropertyLoader.loadProperties(persistenceFile);

        generateIfAbsent(MessagingPropertyKeys.CHANNELID);
        generateIfAbsent(MessagingPropertyKeys.RECEIVERID);

        persistProperties();
    }

    public Properties getPersistedProperties() {
        if (storage == null) {
            return new Properties();
        }
        Properties propertiesCopy = new Properties();
        propertiesCopy.putAll(storage);
        return propertiesCopy;
    }

    private void persistProperties() {
        FileOutputStream fileOutputStream = null;
        try {
            File parentDirectory = persistenceFile.getParentFile();
            if (parentDirectory != null && !parentDirectory.exists() && !parentDirectory.mkdirs()) {
                throw new IllegalStateException("Could not create parent directory: " + parentDirectory);
            }

            if (!persistenceFile.exists()) {
                boolean created = persistenceFile.createNewFile();
                if (!created) {
                    throw new IllegalStateException("Could not create persistenceFile: " + persistenceFile);
                }
            }

            fileOutputStream = new FileOutputStream(persistenceFile);
            storage.store(fileOutputStream, null);
        } catch (IOException e1) {
            logger.error("Couldn't write properties file {}", persistenceFile.getPath(), e1);
        } finally {
            if (fileOutputStream != null) {
                try {
                    fileOutputStream.close();
                } catch (IOException e) {
                }
            }
        }
    }

    private void makeDirectories(File persistenceFile) {
        File parentFile = persistenceFile.getParentFile();
        if (parentFile != null && parentFile.exists() == false) {
            boolean madeDirs = parentFile.mkdirs();
            if (!madeDirs) {
                throw new IllegalStateException("Couldn't create dir: " + parentFile);
            }
        }
    }

    private void generateIfAbsent(String propertyName) {
        if (!storage.containsKey(propertyName)) {
            String propertyValue = createUuidString();
            storage.put(propertyName, propertyValue);
        }
    }
}
